packages/actor-rdf-metadata-extract-void/lib/ActorRdfMetadataExtractVoid.ts

```diff
@@ -0,0 +1,547 @@
+    return Number.parseInt(bindings.at(0)?.get('distinctSubjects')?.value ?? '0', 10);
+  }
+
+  public async getPredicateObjects(dataset: IVoidDataset, predicate: RDF.Term): Promise<number> {
+    const query = `
+      PREFIX void: <http://rdfs.org/ns/void#>
+
+      SELECT ?distinctObjects WHERE {
+        ${termToString(dataset.identifier)} void:propertyPartition [
+          void:property ${termToString(predicate)} ;
+          void:distinctObjects ?distinctObjects
+        ] .
+      } LIMIT 1
+    `;
+    const bindings = await this.getBindings(dataset, query);
+    return Number.parseInt(bindings.at(0)?.get('distinctObjects')?.value ?? '0', 10);
+  }
+
+  public async getClassPartitionEntities(dataset: IVoidDataset, object: RDF.Term): Promise<number> {
+    const query = `
+      PREFIX void: <http://rdfs.org/ns/void#>
+
+      SELECT ?entities WHERE {
+        ${termToString(dataset.identifier)} void:classPartition [
+          void:class ${termToString(object)} ;
+          void:entities ?entities
+        ] .
+      } LIMIT 1
+    `;
+    const bindings = await this.getBindings(dataset, query);
+    return Number.parseInt(bindings.at(0)?.get('entities')?.value ?? '0', 10);
+  }
+
+  public async getBindings(dataset: IVoidDataset, query: string): Promise<RDF.Bindings[]> {
+    let bindings = this.bindingsCache.get(query);
+    if (!bindings) {
+      const bindingsStream = await this.queryEngine.queryBindings(query, { sources: [ dataset.store ]});
+      bindings = await bindingsStream.toArray();
+      this.bindingsCache.set(query, bindings);
+    }
+    return bindings;
+  }
+}
+
+export interface IActorRdfMetadataExtractVoidArgs extends IActorRdfMetadataExtractArgs {
+  /**
+   * An init query actor that is used to query shapes.
+   * @default {<urn:comunica:default:init/actors#query>}
+   */
+  actorInitQuery: ActorInitQueryBase;
+  /**
+   * Whether URI regex patterns should be inferred based on dataset URI or
+   * void:uriSpace if not present in the VoID description.
+   * @default {true}
+   */
+  inferUriRegexPattern: boolean;
+  /**
+   * How many bindings results used within the cardinality estimation formulae should be cached.
+   * This is done to avoid repeated queries ovet the VoID metadata to extract the same handful of numbers.
+   * @default {100}
+   */
+  bindingsCacheSize: number;
+}
+
+/**
+ * Comunica-specific wrapper for the VoID dataset description, as defined in the specification:
+ * https://www.w3.org/TR/void/
+ */
+export interface IVoidDataset {
+  /**
+   * The source IRI of the dataset description.
+   */
+  source: string;
+  /**
+   * The identifier of the dataset, either an IRI or a blank node.
+   */
+  identifier: RDF.NamedNode | RDF.BlankNode;
+  /**
+   * The RDF/JS store containing the data for this dataset description.
+   * Note that the store might also contain data for other datasets defined in the dame VoID description.
+   */
+  store: RDF.Store;
```

Comment:
Do we need to expose this?
Might be safer to hide it wrt future breaking changes?

Reply:
That is a great point! I did not realise it can be hidden. Thank you. :slightly_smiling_face: